### PR TITLE
Initial non-functional UI for overall ratings

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-frontend/issues/1017

This is a first step toward the UI that will let you select an overall rating. It's not functional yet because we need some API features.

screenshot 2016-09-08 15 17 19

### CODE REVIEW

File: src/amo/css/AddonDetail.scss [hunk 1]
```diff
@@ -68,7 +68,7 @@
     }
   }
 
-  .about {
+  .about, .overall-rating {
```
Comment (muffinresearch) [inline src/amo/css/AddonDetail.scss]: Generally we put selectors on a new line (maybe stylelint has a rule for this) because if you have a lot of selectors it's quite hard to parse them reading across the page as opposed to vertically.

File: src/amo/css/OverallRating.scss [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      text-align: center;
+      padding-top: 90px;
+      width: 120px;
+      height: 100px;
+      background-size: 85px;
+      background-position: center;
+      background-repeat: no-repeat;
+    }
+
+    label.love-it {
```
Comment (muffinresearch) [inline src/amo/css/OverallRating.scss]: You can save some repetition here and nest under label.

File: src/amo/components/OverallRating.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      <div className="OverallRating">
+        <p>{prompt}</p>
+        <form action="">
+          <input id="love-it" value="love-it" name="overall-rating" type="radio" />
+          <label htmlFor="love-it" className="love-it">
+            {i18n.gettext('Love it!')}
+          </label>
+          <input id="it-is-ok" value="it-is-ok" name="overall-rating" type="radio" />
+          <label htmlFor="it-is-ok" className="it-is-ok">
+            {i18n.gettext('It\'s OK')}
```
Comment (muffinresearch) [inline src/amo/components/OverallRating.js]: You should be ok to use `"` here to save on the escaping the `;`.

File: src/amo/components/AddonDetail.js [hunk 4]
```diff
@@ -90,6 +91,13 @@ class AddonDetail extends React.Component {
           <div dangerouslySetInnerHTML={sanitizeHTML(nl2br(addon.description),
                                                      allowedDescriptionTags)} />
         </section>
+
+        <hr />
```
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: I used `hr` in the detail page as a quick fix, but since it's presentational markup we should reconsider keeping hrs around.

File: src/amo/components/OverallRating.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          </label>
+          <input id="huh" value="huh" name="overall-rating" type="radio" />
+          <label htmlFor="huh" className="huh">
+            {i18n.gettext('Huh?')}
```
Comment (muffinresearch) [inline src/amo/components/OverallRating.js]: I see we've lost 'meh' but `huh?` seems similarly vague and might be hard to translate. 

We can deal with this as a separate question to UX/Product people later though.

File: src/amo/css/AddonDetail.scss (hunk 1 above)
Comment (muffinresearch) [inline src/amo/css/AddonDetail.scss]: In fact, I missed it - stylelint is already telling you this :)

File: src/amo/css/OverallRating.scss [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      background-image: url('../img/ratings/smiley-love-it.svg');
+    }
+
+    label.it-is-ok {
+      background-image: url('../img/ratings/smiley-it-is-ok.svg');
+    }
+
+    label.huh {
+      background-image: url('../img/ratings/smiley-huh.svg');
```
Comment (muffinresearch) [inline src/amo/css/OverallRating.scss]: We should file something on this to look at using svgo later to crush the svgs (probably something we can do via webpack and also we might want to consider these 3 being a single sprite with something like svg_sprite too although if they're small enough to be data/uris then we won't need to.

File: src/amo/components/OverallRating.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    return (
+      <div className="OverallRating">
+        <p>{prompt}</p>
+        <form action="">
+          <input id="love-it" value="love-it" name="overall-rating" type="radio" />
+          <label htmlFor="love-it" className="love-it">
+            {i18n.gettext('Love it!')}
+          </label>
+          <input id="it-is-ok" value="it-is-ok" name="overall-rating" type="radio" />
```
Comment (mstriemer) [inline src/amo/components/OverallRating.js]: I don't see these `id`s being used. Perhaps they shouldn't be set?

Comment (mstriemer) [inline src/amo/components/OverallRating.js]: Oh, for the label. Can you wrap the input instead or does that mess up the styles?

File: src/amo/css/OverallRating.scss [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    justify-content: center;
+
+    input {
+      appearance: none;
+      opacity: 0;
+      width: 0;
+      height: 0;
+    }
+
+    label {
```
Comment (mstriemer) [inline src/amo/css/OverallRating.scss]: We don't have a specific CSS style guide but I've been trying to keep things as single class names if possible. I think in this case the style engine would need to check each `<label>` to see if it's in a `<form>` and a `.OverallRating`. Probably isn't a big performance hit but if we keep using bare element selectors we could end up with a bunch of that.

I'd prefer:

```scss
.OverallRating-form { }
.OverallRating-input { }
.OverallRating-label { }
.OverallRating-label-love-it { }
// ...
```

It's a bit weird and I'm not sure I'd suggest it but it looks like you can omit the common class portion with:

```scss
.OverallRating {
  &-form { }
  &-input { }
}

**Comment by mstriemer** [comment]
Looks good!

The localisation is definitely going to be tricky. Might be worth adding a comment to them stating that the strings should be really short.

r+wc

File: src/amo/css/OverallRating.scss (hunk 6 above)
Comment (muffinresearch) [inline src/amo/css/OverallRating.scss]: (OT) Also I see we're starting to enable http2 on the CDN, we should start thinking about what changes we'll need to do there as things like spriting aren't required.

File: src/amo/components/AddonDetail.js (hunk 4 above)
Comment (kumar303) [inline src/amo/components/AddonDetail.js]: ah, right. The mock has them so I guess we could use `border-bottom` or something?

File: src/amo/components/OverallRating.js (hunk 7 above)
Comment (kumar303) [inline src/amo/components/OverallRating.js]: I couldn't figure out a way to wrap the input with a background image and also hide the form control graphic. It seems like wrapping the label is common practice and I think it is semantic enough since, in this case, the real label (e.g. "I love it!") is a major piece of the UI

File: src/amo/css/OverallRating.scss (hunk 8 above)
Comment (kumar303) [inline src/amo/css/OverallRating.scss]: doesn't my initial `.OverallRating {}` block accomplish the same thing though? The style engine will only be looking at the tree of elements starting at `<div class="OverallRating" />`

File: src/amo/css/AddonDetail.scss (hunk 1 above)
Comment (kumar303) [inline src/amo/css/AddonDetail.scss]: Hooray for robots

File: src/amo/css/OverallRating.scss (hunk 8 above)
Comment (mstriemer) [inline src/amo/css/OverallRating.scss]: The browser applies styles from right to left. So any time it encounters a `<label>` it will need to start walking up the tree to figure out if it has a parent `<form>` and that form has a parent `.OverallRating`. With `.OverallRating-label` it just looks at the class and is done. Mostly the issue would be with elements that don't match, where it would need to look all the way up to `<body>` to see that there's no `.OverallRating`.

I don't think there's any sort of optimization to limit based on the class at the left of the selector.

Comment (kumar303) [inline src/amo/css/OverallRating.scss]: Is this really worth optimizing for? It makes the code for both the css and HTML very verbose.

Comment (mstriemer) [inline src/amo/css/OverallRating.scss]: I'm not too concerned about the performance I mostly think that using a single class name for selectors is just nicer. It makes specificity much simpler and doesn't have that performance issue.

If we always use a single class name then you don't need to think about it. To figure out the selector for the `&.huh` on line 34 you need to read all the way back to line 1 to figure it out. It appears to be `.OverallRating form label.huh`. I just think that writing `.OverallRating-label-love-it` would be nicer and has other advantages.

If you really want to nest the styles to type less you could but being able to look at the class name and know what it does is pretty nice.

Comment (mstriemer) [inline src/amo/css/OverallRating.scss]: As for the verbosity in the HTML you could make a stateless component and shorten the HTML instead:

```jsx
const OverallRatingChoice = ({ children, value }) => (
  <div>
    <input id="love-it" value="love-it" name="overall-rating"
           className={`OverallRating-label-${value}`} type="radio" />
    <label htmlFor="love-it" className="love-it">{children}</label>
  </div>
);

<OverallRatingChoice value="love-it">{i18n.gettext('Love it!')}</OverallRatingChoice>
<OverallRatingChoice value="it-is-okay">{i18n.gettext("It's okay")}</OverallRatingChoice>
<OverallRatingChoice value="huh">{i18n.gettext('Huh?')}</OverallRatingChoice>

File: src/amo/components/OverallRating.js (hunk 7 above)
Comment (muffinresearch) [inline src/amo/components/OverallRating.js]: Nothing wrong with making the relationship between label and input explicit like this. I've always personally preferred using attributes instead of relying on the implied relationship even when wrapping the input with a label.

Comment (kumar303) [inline src/amo/components/OverallRating.js]: Oh, sorry, I totally misunderstood. I forgot about the possibility of nesting an input inside a label. I just looked up the pros and cons right now and I saw a reminder that you are required to declare `for` on a label to address a11y so I don't think it would buy much by nesting here.

File: src/amo/css/OverallRating.scss (hunk 8 above)
Comment (kumar303) [inline src/amo/css/OverallRating.scss]: oh, is this about label namespacing, like, in case we introduce another component elsewhere for "love it" ? I could see that being a problem. I'll prefix it.

**Comment by kumar303** [comment]
I added a note for L10n and fixed up the class prefixing. @mstriemer thanks for the review!
